server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java

```diff
@@ -144,111 +145,104 @@ public RepositoriesService(
-                publicationStep.onFailure(e);
-                super.onFailure(e);
-            }
+            .<RegisterRepositoryTaskResult>andThen((clusterUpdateStep, ignored) -> {
+                final ListenableFuture<AcknowledgedResponse> acknowledgementStep = new ListenableFuture<>();
+                final ListenableFuture<Boolean> publicationStep = new ListenableFuture<>(); // Boolean==changed.
+                submitUnbatchedTask(
+                    "put_repository [" + request.name() + "]",
+                    new RegisterRepositoryTask(this, request, acknowledgementStep) {
+                        @Override
+                        public void onFailure(Exception e) {
+                            logger.warn(() -> "failed to create repository [" + request.name() + "]", e);
+                            publicationStep.onFailure(e);
+                            super.onFailure(e);
+                        }
 
-            @Override
-            public boolean mustAck(DiscoveryNode discoveryNode) {
-                // repository is created on both master and data nodes
-                return discoveryNode.isMasterNode() || discoveryNode.canContainData();
-            }
+                        @Override
+                        public boolean mustAck(DiscoveryNode discoveryNode) {
+                            // repository is created on both master and data nodes
+                            return discoveryNode.isMasterNode() || discoveryNode.canContainData();
+                        }
 
-            @Override
-            public void clusterStateProcessed(ClusterState oldState, ClusterState newState) {
-                if (changed) {
-                    if (found) {
-                        logger.info("updated repository [{}]", request.name());
-                    } else {
-                        logger.info("put repository [{}]", request.name());
+                        @Override
+                        public void clusterStateProcessed(ClusterState oldState, ClusterState newState) {
+                            if (changed) {
+                                if (found) {
+                                    logger.info("updated repository [{}]", request.name());
+                                } else {
+                                    logger.info("put repository [{}]", request.name());
+                                }
+                            }
+                            publicationStep.onResponse(oldState != newState);
+                        }
                     }
+                );
+                publicationStep.addListener(clusterUpdateStep.delegateFailureAndWrap((ignored1, changed) -> {
+                    acknowledgementStep.addListener(clusterUpdateStep.delegateFailureAndWrap((ignored2, ack) -> {
+                        clusterUpdateStep.onResponse(new RegisterRepositoryTaskResult(ack, changed));
+                    }));
+                }));
```

Comment:
Best not to capture the listener in the lambdas (probably doesn't matter here, we're capturing other things, but still good practice). Also `delegateFailure(onResponse(...))` is just `map(...)` (the `AndWrap` bit shouldn't matter here).

```suggestion
                publicationStep.addListener(
                    clusterUpdateStep.delegateFailureAndWrap(
                        (clusterUpdateStep1, changed) -> acknowledgementStep.addListener(
                            clusterUpdateStep1.map(ack -> new RegisterRepositoryTaskResult(ack, changed))
                        )
                    )
                );
```